presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SingleDistinctOptimizer.java

```diff
@@ -0,0 +1,150 @@
+                                        Collections.emptyMap(),
+                                        node.getStep(),
+                                        node.getSampleWeight(),
+                                        node.getConfidence(),
+                                        node.getHashSymbol());
+        }
+
+        boolean canOptimize(AggregationNode node)
+        {
+            // all aggregation functions have a single common distinct mask symbol
+            // AND all aggregation functions have mask
+            return node.getMasks().values().stream().distinct().count() == 1 &&
+                    node.getMasks().size() == node.getAggregations().size();
+        }
+
+        public PlanNode visitMarkDistinct(MarkDistinctNode node, RewriteContext<Optional<Symbol>> context)
+        {
+            Optional<Symbol> mask = context.get();
+            if (mask.isPresent() && mask.get().equals(node.getMarkerSymbol())) {
+                // rewrite Distinct into GroupBy
+                AggregationNode aggregationNode = new AggregationNode(idAllocator.getNextId(),
+                                                                        node.getSource(),
+                                                                        node.getDistinctSymbols(),
+                                                                        Collections.emptyMap(),
+                                                                        Collections.emptyMap(),
+                                                                        Collections.emptyMap(),
+                                                                        SINGLE,
+                                                                        Optional.empty(),
+                                                                        1.0,
+                                                                        node.getHashSymbol());
+
+                ImmutableMap.Builder<Symbol, Expression> outputSymbols = ImmutableMap.builder();
+                for (Symbol symbol : node.getDistinctSymbols()) {
+                    Expression expression = new QualifiedNameReference(symbol.toQualifiedName());
+                    outputSymbols.put(symbol, expression);
+                }
+
+                if (aggregationNode.getHashSymbol().isPresent()) {
+                    Expression hashExpression = getHashExpression(node.getDistinctSymbols());
+                    outputSymbols.put(symbolAllocator.newHashSymbol(), hashExpression);
+                }
+                // add null assignment for mask
+                // unused mask will be removed by PruneUnreferencedOutputs
+                outputSymbols.put(mask.get(), new NullLiteral());
+                return new ProjectNode(idAllocator.getNextId(),
+                                        aggregationNode,
+                                        outputSymbols.build());
+            }
+            return node;
```

Comment:
This should invoke context.defaultRewrite() so that the optimization can apply recursively.